### PR TITLE
Config v6

### PR DESCRIPTION
**Pull request by novalisdenahi**
### Describe the purpose of your pull request

Features/improvements:
* [x] Fetch new config json version (`config_v6.json`) and update config model
* [x] SDK key format validation (when client is not set up to use local-only flag overrides)
* [x] AND conditions
* [x] New condition types
   * Prerequisite flag condition
   * Segment condition
* [x] Changed comparison operators
   * (NOT) CONTAINS ANY OF
* [x] New comparison operators
   * BEFORE/AFTER
   * TEXT (NOT) EQUALS (both hashed and cleartext)
   * TEXT (NOT) STARTS/ENDS WITH ANY OF (both hashed and cleartext)
   * ARRAY CONTAINS ANY OF (both hashed and cleartext)
* [x] Percentage options within targeting rules
* [x] Percentage options based on custom attributes
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] User Object attribute/comparison value trimming
* [x] Comparison value hashing changes (SHA256, salting)
* [ ] ~Flag overrides vs. config json salt and segments~ (No Json override option)
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Update `ConfigCatClient.getKeyAndValue`
* [ ] Setting DataSource change to Config Data source

Tests: 
* [x] Update cache key/payload tests
* [x] Evaluation matrix tests
  * Config v1 features (both V1 & V2)
  * Config v2 features
* [x] Evaluation log tests
* [x] SDK key format validation
* [x] Custom User Object attribute handling (don't force string values + autoconversions under the hood)
* [x] Flag overrides vs. prerequisite flags (circular dependency, comparison value type mismatch, dependent/prereq flag overriding)
* [ ] ~Flag overrides vs. config json salt and segments~ (No Json override option)
* [x] `EvaluationDetails.matchedTargetingRule`/`matchedPercentageOption` properties (rename + set combinations correctly)
* [x] Special character tests (https://trello.com/c/qh4jHrvm)
* [x] Comparison attribute and comparison value trimming tests (https://trello.com/c/vt8FFrSZ)
* [x] Comparison attribute conversion to canonical string representation tests (https://trello.com/c/CZgGrrcb)

Related improvements/fixes:
* [ ] ~Dynamically typed languages should log a warning when default value doesn't match the setting type (see https://trello.com/c/Le6vimGu)~
* [ ] ~We should go to the cache in all polling modes instead of using the in memory variable (see https://trello.com/c/rreKm64A)~ (on an other PR, released together)
* [x] Improve message of log event 1103 (https://trello.com/c/Mb7rvm64)

Breaking changes:

- Rework `Config` and related ConfigCat models to support v6.
- Rename `EvaluationDetails` matchedEvaluationRule` to `matchedTargetingRule` and `matchedEvaluationPercentageRule` to `matchedPercentageOption`.
- Setting DataSource replaced with Config DataSource

### Requirement checklist (only if applicable)

- [x] I have covered the applied changes with automated tests.
- [x] I have executed the full automated test set against my changes.
- [x] I have validated my changes against all supported platform versions.
- [x] I have read and accepted the [contribution agreement](https://github.com/configcat/legal/blob/main/contribution-agreement.md).

### CODE REVIEW

**Review by adams85 — CHANGES_REQUESTED** [review]
Apart from the comments, I have some further tangential remarks:
1. As it turned out, the Kotlin SDK doesn't have a real end-to-end test (like a `getValue` call which pulls the config data from the CDN) as the HTTP communication is mocked in all the tests. That is, we can't really call e.g. the matrix tests integration tests currently. Obviously, this situation is not good and I think we should change it.
2. Another issue with the matrix tests and other data file-driven tests that the test data is not read from file but hardcoded in source files. This makes the review process harder and less efficient and also hurts maintainability. (For example, when we need to alter the test data in the future, a simple file copy won't be possible unlike in other SDKs.)
3. Exceptions used to control flow in the evaluation logic. (The same issue as the one discussed here: https://github.com/configcat/java-sdk/pull/39#discussion_r1439352583)

Based on our discussion with @novalisdenahi , solving these issues are not trivial and may take longer time. So I don't insist on solving them now, but I think we should take care of them sooner or later. As a matter of fact, I wouldn't mind if issue 1 were addressed now, but the decision is not up to me.

File: src/commonMain/kotlin/com/configcat/Evaluator.kt [hunk 1]
```diff
@@ -1,437 +1,1169 @@
...
         setting: Setting,
-        user: ConfigCatUser,
-        evaluatorLogger: EvaluatorLogger
+        context: EvaluationContext,
+        evaluateLogger: EvaluateLogger?
     ): EvaluationResult? {
-        if (setting.rolloutRules.isEmpty()) {
+        if (setting.targetingRules.isNullOrEmpty()) {
             return null
         }
```
[the remark this answers is not in the document]
Comment (novalisdenahi) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt:115]: We have to check again in the method, or we could use the ! operator, but we don't want that.

File: src/commonMain/kotlin/com/configcat/Evaluator.kt [hunk 2]
```diff
@@ -1,437 +1,1169 @@
...
-                    userValue ?: "",
-                    comparator,
-                    rule.comparisonValue
-                )
+
+            if (rule.servedValue != null) {
+                return EvaluationResult(rule.servedValue.value, rule.servedValue.variationId, rule, null)
+            }
+            if (rule.percentageOptions.isNullOrEmpty()) {
                 continue
```
[the remark this answers is not in the document]
Comment (novalisdenahi) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt:145]: Do you mean the error throw should be in the `getKeyAndValue`? The check is there `targetingRule.percentageOptions?` but we didn't throw error in the other SDKs as well. I don't think it's necessary.

File: src/commonMain/kotlin/com/configcat/Evaluator.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (adams85) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt:115]: How about passing the array of targeting rules to the `evaluateTargetingRules` method as a non-nullable parameter? Something like this:

```kotlin
  private fun evaluateTargetingRules(
        targetingRules: Array<TargetingRule>,
        /*...*/
```

File: src/commonMain/kotlin/com/configcat/Evaluator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (adams85) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt]: As you can see, [the reference implementation does throw an error](https://github.com/configcat/.net-sdk/blob/985600c2f91c1b388187d09e50a0d3af40cf6bdc/src/ConfigCatClient/Evaluation/RolloutEvaluator.cs#L156), while the Kotlin SDK just continues evaluation (if I'm not missing here something).

As for `getKeyAndValue`, .NET SDK doesn't have this method currently, but others check this case:
* https://github.com/configcat/php-sdk/blob/master/src/ConfigCatClient.php#L546 (`hasPercentageOptions` throws when the THEN part is missing)
* https://github.com/configcat/cpp-sdk/blob/a6adaa0b56d7579448ac788fa83bc4c053d3380a/src/configcatclient.cpp#L317-L319
* https://github.com/configcat/swift-sdk/blob/c748d1bab8fc64f6b11cba9f131e5fbd789ed3b9/Sources/ConfigCat/ConfigCatClient.swift#L298-L300

If other SDKs don't do this check, that's an oversight and slipped past review unfortunately.

File: src/commonMain/kotlin/com/configcat/Evaluator.kt [hunk 3]
```diff
@@ -1,437 +1,1169 @@
...
+        try {
+            if (userAttribute is Array<*> && userAttribute.all { it is String }) {
+                return userAttribute as Array<String>
+            }
+            if ((userAttribute is List<*>) && userAttribute.all { it is String }) {
+                val stringList: List<String> = userAttribute as List<String>
+                return stringList.toTypedArray()
+            }
+            if (userAttribute is String) {
+                return Constants.json.decodeFromString(userAttribute)
```
[the remark this answers is not in the document]
Comment (novalisdenahi) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt:865]: It shouldn't allow null; the method returns with Array<String>, not with Array<String?>

File: src/commonMain/kotlin/com/configcat/Evaluator.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (novalisdenahi) [reply src/commonMain/kotlin/com/configcat/Evaluator.kt]: Yep, this should be fixed in Java, Android and Dart as well.

File: src/commonMain/kotlin/com/configcat/model/SettingValue.kt [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    @SerialName("s")
+    var stringValue: String? = null,
+
+    @SerialName("i")
+    var integerValue: Int? = null,
+
+    @SerialName("d")
+    var doubleValue: Double? = null
+) {
+    public fun equalsBasedOnSettingType(other: Any?, settingType: Int): Boolean {
```
Comment (adams85) [inline src/commonMain/kotlin/com/configcat/model/SettingValue.kt:24]: This method serves a special internal purpose, I'd rather not expose it to the end users.

```suggestion
    internal fun equalsBasedOnSettingType(other: Any?, settingType: Int): Boolean {
```

File: src/commonMain/kotlin/com/configcat/model/SettingValue.kt [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+            SettingType.INT -> {
+                integerValue == other.integerValue
+            }
+
+            SettingType.DOUBLE -> {
+                doubleValue == other.doubleValue
+            }
+
+            SettingType.JS_NUMBER -> {
+                doubleValue == other.doubleValue || integerValue == other.integerValue
```
Comment (adams85) [inline src/commonMain/kotlin/com/configcat/model/SettingValue.kt:48]: Is it safe to assume that a value like `1` is always stored in `integerValue`, never in `doubleValue` (as `1.0`)?

How about something like this:

```suggestion
                (doubleValue ?? integerValue.toDouble()) == (other.doubleValue ?? other.integerValue.toDouble())
```

File: src/commonMain/kotlin/com/configcat/Evaluator.kt [hunk 6]
```diff
@@ -1,437 +1,1171 @@
...
+        }
+        if (userValue is Array<*> && userValue.all { it is String }) {
+            return Constants.json.encodeToString(userValue as Array<String>)
+        }
+        if (userValue is List<*> && userValue.all { it is String }) {
+            return Constants.json.encodeToString(userValue as List<String>)
+        }
+        if (userValue is Float) {
+            // toSting and toDouble hack needed in case of float. else the 0.12345 is converted to something like 0.1234500120130
+            return doubleToString(userValue.toString().toDouble())
```
Comment (adams85) [inline src/commonMain/kotlin/com/configcat/Evaluator.kt:1017]: I checked the reference implementation and it also does the "0.12345 is converted to something like 0.1234500120130" thing.

This is how float to double conversion works according to the IEEE 754 standard, so let's say that this is by design and remove the hack.

**Review by adams85 — CHANGES_REQUESTED** [review]

File: src/commonMain/kotlin/com/configcat/ConfigCatUser.kt [hunk 7]
```diff
@@ -1,41 +1,79 @@
...
+ *
+ * Custom attributes of the user for advanced targeting rule definitions (e.g. user role, subscription type, etc.)
+ *
+ * The set of allowed attribute values depends on the comparison type of the condition which references the User Object attribute.<br>
+ * [String] values are supported by all comparison types (in some cases they need to be provided in a specific format though).<br>
+ * Some of the comparison types work with other types of values, as described below.
+ *
+ * Text-based comparisons (EQUALS, IS ONE OF, etc.)<br>
+ *  * accept [String] values,
+ *  * all other values are automatically converted to string (a warning will be logged but evaluation will continue as normal).
```
Comment (adams85) [inline src/commonMain/kotlin/com/configcat/ConfigCatUser.kt:17]:
```suggestion
 *  * all other values are automatically converted to `String` (a warning will be logged but evaluation will continue as normal).
```

**Review by adams85 — APPROVED** [review]

File: src/jsMain/kotlin/com/configcat/fetch/HttpRequestBuilder.kt [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package com.configcat.fetch
+
+import io.ktor.client.request.*
+
+internal actual fun httpRequestBuilder(
+    configCatUserAgent: String,
+    eTag: String
+): HttpRequestBuilder {
+    val httpRequestBuilder = HttpRequestBuilder()
```
Comment (z4kn4fein) [inline src/jsMain/kotlin/com/configcat/fetch/HttpRequestBuilder.kt:9]: ❤️

**Review by z4kn4fein — APPROVED** [review]
Please fix the reported detect warnings, but other than that 💚

**Pull request merged**